lint: run goheader enforcing standardized Go header template

ti-mo opened the pull request:
Since we didn't have a linter to enforce a particular Go file header format, this patch adds one with a relatively simple structure.

goheader can only match on a single block of // statements, so anything that cannot be encoded in the template needed to be moved to a separate block.

For Hubble packages, 'Authors of Hubble' was kept as the primary tagline. Any other copyright notices were moved to subsequent blocks.

Fixes: https://github.com/cilium/cilium/pull/20769#issuecomment-1281718208

```release-note
Fix up and lint SPDX headers in all Go files
```

ti-mo committed:
pkg/bpf: sync copyright headers with package binary

These files were mostly taken from the Go stdlib. It seems like the Go project
has also moved to a lean file header pointing to a single LICENSE file in the
project. Retain only the relevant Go Authors copyright notice and point to the
original files.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
*: add missing SPDX headers to all Go files

Signed-off-by: Timo Beckers <timo@isovalent.com>

sayboras reviewed (approved):
:astonished:

joestringer commented on .golangci.yaml:
```diff
@@ -49,17 +49,19 @@ linters-settings:
       - nilness
   goimports:
     local-prefixes: github.com/cilium/cilium
-  misspell:
-    # Correct spellings using locale preferences for US or UK.
-    # Default is to use a neutral variety of English.
-    # Setting locale to US will correct the British spelling of 'colour' to 'color'.
-    # locale: US
-    ignore-words:
-      - rela # This is for elf.SHT_RELA
```
Was this change intentional?

ti-mo replied:
It was. iirc the code this was necessary for was removed with the alignchecker rewrite. Does this need to go into a separate commit?

joestringer replied:
Ideally yeah, in particular I asked the question because the commit message didn't mention it at all, so I wasn't sure how it related to the linting changes.

ti-mo replied:
Sure thing! :+1:

joestringer commented on pkg/bpf/binary/binary.go:
```diff
@@ -1,33 +1,7 @@
-//
-// Redistribution and use in source and binary forms, with or without
-// modification, are permitted provided that the following conditions are
-// met:
```
I think the intent behind the previous arrangement for this code was that we had relicensed this code Apache-2.0, but it is still subject to the original license. Given the wording here, I don't think we can remove this licensing text.

ti-mo replied:
Ah, so it was relicensed after all; I'll change it back. Didn't know if Apache2 and BSD licences were compatible.

The licensing text has been removed in correspondence to the upstream file, see commit msg. The full headers are no longer included.

joestringer replied:
Ah I see, in that case I think the changes here are fine. Could you also link the upstream location in the commit message so it's easier to verify this?

ti-mo replied:
Sure thing! Commit messages are not documentation, so I've added a reference to the path of the original files in the comments.

joestringer reviewed (approved):
Based on the discussions on my prior feedback, I don't think any changes to the content are necessary. It'd be nice to update the commit messages to highlight the background to the changes but otherwise this LGTM. Thanks for this, it's caught a bunch of missing licenses in source files :tada:

ti-mo committed:
lint: run goheader enforcing standardized Go header template

Since we didn't have a linter to enforce a particular Go file header format,
this patch adds one with a relatively simple structure.

goheader can only match on a single block of // statements, so anything that
cannot be encoded in the template needed to be moved to a separate block.

For Hubble packages, 'Authors of Hubble' was kept as the primary tagline.
Any other copyright notices were moved to subsequent blocks.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
ci: remove misspell linter config for ignoring SHT_RELA

ca451eb5bf removed the use of SHT_REL(A), making this unnecessary.

Signed-off-by: Timo Beckers <timo@isovalent.com>
